test/src/utils/kalman_multivariate.test.cpp

```diff
@@ -0,0 +1,230 @@
+  MatrixXf Q = MatrixXf::Random(n, n);
+  MatrixXf H = MatrixXf::Random(m, n);
+  MatrixXf R = MatrixXf::Random(m, m);
+  MatrixXf P = MatrixXf::Random(n, n);
+
+  void SetUp()
+  {
+    // Assigns the matrices and vectors to the properties within the kalman object.
+    kalman.setModels(A, Q, H, R);
+    kalman.setInitial(x0, P);
+  }
+  void TearDown() {}
+};
+
+/**
+ * Test fixture used for determining whether Kalman handles the zero vector as input for the state
+ * estimates. Should accept
+ */
+TEST_F(KalmanFunctionality, handlesZeroStateEstimate)
+{
+  ASSERT_EQ(kalman.getStateEstimate(), x0) << "Should handle zero vector as state estimate";
+}
+
+/**
+ * Test fixture used for determining whether Kalman handles an arbitrary vector (using Random()) as
+ * input for the state estimates. This should always be accepted.
+ */
+TEST_F(KalmanFunctionality, handlesArbitraryStateEstimate)
+{
+  kalman.setInitial(x1, P);
+  ASSERT_EQ(kalman.getStateEstimate(), x1) << "Should handle any arbitrary state estimate";
+}
+
+/**
+ * Test fixture used for determining whether Kalman handles the zero matrix as input for the state
+ * covariance. This should always be accepted.
+ */
+TEST_F(KalmanFunctionality, handlesZeroStateCovariance)
+{
+  kalman.setInitial(x0, P0);
+  ASSERT_EQ(kalman.getStateCovariance(), P0) << "Should handle zero state covariance";
+}
+
+/**
+ * Test fixture used for determining whether Kalman handles an arbitrary matrix (using Random()) as
+ * input for the state covariance. This should always be accepted.
+ */
+TEST_F(KalmanFunctionality, handlesArbitraryStateCovariance)
+{
+  ASSERT_EQ(kalman.getStateCovariance(), P) << "Should handle arbitrary state covariance";
+}
+
+// -------------------------------------------------------------------------------------------------
+// Mathematical Properties
+// -------------------------------------------------------------------------------------------------
+
+/**
+ * Struct used for testing the mathematical properties of kalman_multivariate. Constructor initiates
+ * an array of Vectors and Matrices to store arbitrary vectors for testing purposes. The arrays are
+ * then populated with values so they can be tested.
+ */
+struct KalmanMathematics : public::testing::Test {
+  unsigned int n = 3;
+  unsigned int m = 2;
+  unsigned int k = 1;
+
+ protected:
+  KalmanMultivariate kalmanMathWithoutControl = KalmanMultivariate(n, m, 0);
+  KalmanMultivariate kalmanMathWithControl = KalmanMultivariate(n, m, k);
+  VectorXf x1_Data[100];
+  VectorXf z_Data[100];
+  VectorXf u_Data[100];
+  MatrixXf A_Data[100];
+  MatrixXf B_Data[100];
+  MatrixXf Q_Data[100];
+  MatrixXf H_Data[100];
+  MatrixXf R_Data[100];
+  MatrixXf P_Data[100];
+  MatrixXf I = MatrixXf::Identity(n, n);
+
+  void SetUp()
+  {
+    // Populates the arrays defined above with 100 random values
+    for (int i = 0; i < 100; i++) {
+      VectorXf x1 = VectorXf::Random(n);
+      x1_Data[i] = x1;
```

Comment:
Why are you not using the following?

```cpp
x1_Data[i] = VectorXf::Random(n);
```

It seems a bit bloated this way. Unless there is a good reason or something in the style guide that I am missing.

Reply:
Yeah thats a much better shout.
Thanks!